craft_parts/parts.py

```diff
@@ -77,6 +78,31 @@ def validate_relative_path_list(cls, item):
         ), f"{item!r} must be a relative path (cannot start with '/')"
         return item
 
+    @root_validator(pre=True)
+    def validate_root(cls, values):
+        """Check if the part spec has a valid configuration of packages and slices."""
+        assert "stage-slices" not in values, (
+            "'stage-slices' must not be explicitly passed in the part data; "
+            "it is derived from 'stage-packages'"
+        )
```

Comment:
It's better to keep the `PartSpec` model reflecting what the user can actually define instead of creating a phantom field as this could cause unexpected behavior when marshaling the part data again. We could use a simple validator for `stage_packages` to make sure it contains only one type of package or other and install slices from `stage_packages` later instead of `stage_slices` if we detect it's using slices.

Reply:
I see. I actually agree with you with maintaining the "purity" of the `PartSpec`, and I considered keeping everything in `stage_packages`. The reason that I chose this path is that if we keep everything in `stage_packages` then every bit of code that references `stage_packages` today will have to be updated because that list might contain slices instead of .deb packages.

So I guess my question is, is that acceptable? Is there code in other repos (like the *craft applications) that manipulate `stage_packages`?

Reply:
The idea is that `stage_packages` could be used regardless of actual content throughout the code except for type-specific code in the `packages` package. Since the package type is set for the entire project and slices can be used per-part (or so it seems), we could handle that in `packages.deb` and (hopefully) leave the rest of the code unchanged.